aiohttp/client_reqrep.py

```diff
@@ -817,505 +549,780 @@ async def close(self) -> None:
+        """Set basic auth."""
+        if auth is None:
+            auth = self.auth
+        if auth is None and trust_env and self.url.host is not None:
+            netrc_obj = netrc_from_env()
+            with contextlib.suppress(LookupError):
+                auth = basicauth_from_netrc(netrc_obj, self.url.host)
+        if auth is None:
+            return
 
-    @reify
-    def request_info(self) -> RequestInfo:
-        return self._request_info
+        if not isinstance(auth, BasicAuth):
+            raise TypeError("BasicAuth() tuple is required instead")
 
-    @reify
-    def content_disposition(self) -> Optional[ContentDisposition]:
-        raw = self._headers.get(hdrs.CONTENT_DISPOSITION)
-        if raw is None:
-            return None
-        disposition_type, params_dct = multipart.parse_content_disposition(raw)
-        params = MappingProxyType(params_dct)
-        filename = multipart.content_disposition_filename(params)
-        return ContentDisposition(disposition_type, params, filename)
+        self.headers[hdrs.AUTHORIZATION] = auth.encode()
 
-    def __del__(self, _warnings: Any = warnings) -> None:
-        if self._closed:
-            return
+    def _update_headers(self, headers: CIMultiDict[str]) -> None:
+        """Update request headers."""
+        self.headers = CIMultiDict[str]()
 
-        if self._connection is not None:
-            self._connection.release()
-            self._cleanup_writer()
+        # Build the host header
+        host = self.url.host_port_subcomponent
 
-            if self._loop.get_debug():
-                _warnings.warn(
-                    f"Unclosed response {self!r}", ResourceWarning, source=self
-                )
-                context = {"client_response": self, "message": "Unclosed response"}
-                if self._source_traceback:
-                    context["source_traceback"] = self._source_traceback
-                self._loop.call_exception_handler(context)
+        # host_port_subcomponent is None when the URL is a relative URL.
+        # but we know we do not have a relative URL here.
+        assert host is not None
+        self.headers[hdrs.HOST] = headers.pop(hdrs.HOST, host)
+        if not self._supports_body:
+            self.headers[hdrs.CONTENT_LENGTH] = "0"
 
-    def __repr__(self) -> str:
-        out = io.StringIO()
-        ascii_encodable_url = str(self.url)
-        if self.reason:
-            ascii_encodable_reason = self.reason.encode(
-                "ascii", "backslashreplace"
-            ).decode("ascii")
-        else:
-            ascii_encodable_reason = "None"
-        print(
-            "<ClientResponse({}) [{} {}]>".format(
-                ascii_encodable_url, self.status, ascii_encodable_reason
-            ),
-            file=out,
-        )
-        print(self.headers, file=out)
-        return out.getvalue()
+        self.headers.extend(headers)
```

Comment:
My assumption is that this does the same thing as the old code, but optimised in Cython. So something seems to be wrong here as the benchmark is coming in much slower.

Reply:
Actually, I think this is faster. For an empty body, I started setting the Content-Length header in this function, so that may be why it's reporting a regression. That code would have been run (likely in a less efficient way) elsewhere in the old code, so doesn't show up in this benchmark.

Reply:
No, wait, ClientRequest won't set that line, the condition would always be False. I'm not sure why it's reporting slower, the codspeed graph just shows that update_headers() is slower, but everything under it is faster.